src/main/java/org/springframework/data/couchbase/repository/support/DynamicInvocationHandler.java

```diff
@@ -0,0 +1,120 @@
+ */
+public class DynamicInvocationHandler<T> implements InvocationHandler {
+	final T target;
+	final Class<?> repositoryClass;
+	final CouchbaseEntityInformation<?, String> entityInformation;
+	final ReactiveCouchbaseTemplate reactiveTemplate;
+	CommonOptions<?> options;
+	String collection;
+	String scope;;
+
+	public DynamicInvocationHandler(T target, CommonOptions<?> options, String collection, String scope) {
+		this.target = target;
+		if (target instanceof CouchbaseRepository) {
+			reactiveTemplate = ((CouchbaseTemplate) ((CouchbaseRepository) target).getOperations()).reactive();
+		} else if (target instanceof ReactiveCouchbaseRepository) {
+			reactiveTemplate = (ReactiveCouchbaseTemplate) ((ReactiveCouchbaseRepository) target).getOperations();
+		} else {
+			throw new RuntimeException("Unknown target type: " + target.getClass());
+		}
+		this.entityInformation = target instanceof CouchbaseRepository
+				? ((CouchbaseRepository<?, String>) target).getEntityInformation()
+				: ((ReactiveCouchbaseRepository<?, String>) target).getEntityInformation();
+		this.options = options;
+		this.collection = collection;
+		this.scope = scope;
+		this.repositoryClass = target.getClass();
+
+	}
+
+	@Override
+	public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+
+		if ("toString".equals(method.getName())) {
+			return "proxy -> target:" + target;
+		}
+		/* Cannot fall-through to use these methods on target, as they will not retain
+		 * the scope, collection and options that may already be set on the proxy
+		 */
+
+		if (method.getName().equals("withOptions")) {
+			return Proxy.newProxyInstance(this.getClass().getClassLoader(), target.getClass().getInterfaces(),
+					new DynamicInvocationHandler<>(target, (CommonOptions) args[0], collection, scope));
+		}
+
+		if (method.getName().equals("withScope")) {
+			return Proxy.newProxyInstance(this.getClass().getClassLoader(), target.getClass().getInterfaces(),
+					new DynamicInvocationHandler<>(target, options, collection, (String) args[0]));
+		}
+
+		if (method.getName().equals("withCollection")) {
+			return Proxy.newProxyInstance(this.getClass().getClassLoader(), target.getClass().getInterfaces(),
+					new DynamicInvocationHandler<>(target, options, (String) args[0], scope));
+		}
+
+		Class<?>[] paramTypes = null;
+		if (args != null) {
+			// the CouchbaseRepository methods - save(entity) etc - will have a parameter type of Object instead of entityType
+			// so change the paramType to match
+			paramTypes = Arrays.stream(args)
+					.map(o -> o == null ? null : (o.getClass() == entityInformation.getJavaType() ? Object.class : o.getClass()))
+					.toArray(Class<?>[]::new);
+		}
+
+		Method theMethod = repositoryClass.getMethod(method.getName(), paramTypes);
+		Object result;
+
+		try {
+			setThreadLocal();
+			result = theMethod.invoke(target, args);
+		} catch (InvocationTargetException ite) {
+			throw ite.getCause();
+		}
+		return result;
+	}
```

Comment:
similar to CrudMethodMetadataPostProcessor - there is no need to reset the thread-local data after the call as this is in the same thread.